lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1673,6 +1674,52 @@ bool TypeChecker::typeCheckFunctionBodyUntil(FuncDecl *FD,
   BraceStmt *BS = FD->getBody();
   assert(BS && "Should have a body");
 
+  // FIXME: FIXME_NOW: DETERMINE: What is the appropriate place for this code to
+  //        live?
+  if (FD->hasSingleExpressionBody()) {
+    auto fnTypeLoc = FD->getBodyResultTypeLoc();
+    auto fnType = fnTypeLoc.getType();
+    auto returnStmt = cast<ReturnStmt>(BS->getElement(0).get<Stmt *>());
+    auto E = returnStmt->getResult();
+
+    // FIXME: FIXME_NOW: DETERMINE: Does this actually need to be here?  It's 
+    //        being done for closures, but there may be a reason that applies
+    //        there but not here.
+    //
+    //           if (E != FD->getSingleExpressionBody()) {
+    //             FD->setSingleExpressionBody(E);
+    //           }
+
+    if (fnTypeLoc.isNull() || fnType->isVoid()) {
+      BS = BraceStmt::create(Context,
+                             BS->getStartLoc(),
+                             { E },
+                             BS->getEndLoc(),
+                             /*implicit=*/false);
+    } else {
+      // FIXME: FIXME_NOW: DETERMINE: Is this the appropriate mechanism to use
+      //        to divine the type of E?
```

Comment:
Not quite, because it might be context-sensitive:

```swift
var foo: String { return .init() }
var bar: Int8 { return 0 }
```

Besides that it seems mostly reasonable, but I'll have to double-check, or get someone more familiar with the type checker to answer.

Reply:
It looks like these two cases work.  You inspired me to add ( 
https://github.com/apple/swift/pull/23251/commits/a5f94069a63f7a144de20e39113a3ec721bea52d ) a lot more tests, though, with functions that implicitly return a number of additional classes of expr.  There are a few implicit exprs for which the implicit return triggers assertion failures: `CoerceExpr`, `ConditionalCheckedCastExpr`, `ForcedCheckedCastExpr`, `IsExpr`, and `ConditionalExpr`: `Assertion failed: (!ifExpr->isFolded() && "already folded if expr in sequence?!")`  If I understand the assertion failure and the effect of `preCheckExpression` correctly, the problem is that `preCheckExpr` folds the sequence and then calling `typeCheckBody` at the end tries to fold it again.  But it looks like `preCheckExpr` has to be called before solving the constraint system.

Reply:
Caching the result (a bit for whether it's been preChecked and a pointer to the replacement expr if it has) of `preCheckExpr` on the Expr being prechecked avoids that assert.  Does that sound like a reasonable solution, @jrose-apple?

```
/// Pre-check the expression, validating any types that occur in the
/// expression and folding sequence expressions.
bool TypeChecker::preCheckExpression(Expr *&expr, DeclContext *dc) {
  if (expr->getIsPreChecked()) {
    auto result = expr->getPreCheckResult();
    expr = result;
    return result == nullptr;
  }
  PreCheckExpression preCheck(*this, dc, expr);
  // Perform the pre-check.
  if (auto result = expr->walk(preCheck)) {
    expr->setPreChecked(result);
    expr = result;
    return false;
  }
  expr->setPreChecked(nullptr);
  return true;
}
```

Reply:
:-( I don't like that we don't know if it's going to happen or not. Can we go the other way instead, where we just say single-expression functions never go down the normal path, and just call typeCheckStatement explicitly if they end up making a new statement? That's closer to what we're doing for closures.

Reply:
I was overlooking the fact that `preCheckExpression` was handing back the modified (potentially folded) expression.  Writing the updated expression back into the single expression body field of the function resolves the refolding assertion.  This feels a lot better to me than what I was proposing above.  

The type-checking does still occur twice, but we can use the expression's type to determine whether or not to remove the return, just like the closure case is doing.

```
// CSApply.cpp:7522
if (cs.getType(body)->isUninhabited()) {
  closure = Rewriter.coerceClosureExprFromNever(closure);
}
```

```
// CSApply.cpp:5864
auto returnStmt = cast<ReturnStmt>(member.get<Stmt *>());
auto singleExpr = returnStmt->getResult();

cs.setExprTypes(singleExpr);
tc.checkIgnoredExpr(singleExpr);

SmallVector<ASTNode, 1> elements;
elements.push_back(singleExpr);

auto braceStmt = BraceStmt::create(tc.Context,
                                   closureExpr->getStartLoc(),
                                   elements,
                                   closureExpr->getEndLoc(),
                                   /*implicit*/true);
```

What do you think about this approach?

Reply:
> The type-checking does still occur twice, but we can use the expression's type to determine whether or not to remove the return, just like the closure case is doing.

I don't understand this bit. In the closure case, we have an extra conversion between the type of the closure literal and the expected type, and that's where the coercion happens. We don't have that here, where you're just type-checking the plain expression (per the other comment with a preferred overload).

Reply:
I was trying to highlight the similarity between this approach and what's being done for closures, but I recognize the difference you describe.  I'll see about changing the implementation to align more closely with the approach taken for closures.